### PR TITLE
Initial version of visualization code for unary relations over a single domain

### PR DESCRIPTION
**Pull request by ThomasColthurst**
Sample output image when run with
$ ./visualize_unary.py --observations=../cxx/assets/animals.unary.obs --clusters=../cxx/assets/animals.unary.hirm --output=/tmp/vis.html

download

### CODE REVIEW

**Comment by emilyfertig** [comment]
Reassigned to @srvasude (who agreed to review while I focus on Model 7)

File: py/hirm_io.py [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+def load_clusters(path):
+  """Load the hirm clusters output from path as an HIRM_Clusters."""
+  id_to_relations = {}
+  id_to_clusters = {}
+
+  with open(path, 'r') as f:
+    # Read the relation clusters
+    for line in f:
+      fields = line.rstrip().split()
+      if len(fields) == 0:
```
[the remark this answers is not in the document]
Comment (ThomasColthurst) [reply py/hirm_io.py]: Done.

File: py/visualize.py [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+
+import base64
+import copy
+import io
+import sys
+import numpy as np
+from matplotlib import pyplot as plt
+
+
+def figure_to_html(fig: plt.Figure) -> str:
```
[the remark this answers is not in the document]
Comment (ThomasColthurst) [reply py/visualize.py]: Removed.

File: py/visualize_unary.py [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+  print(f"Loading observations from {args.observations} ...")
+  obs = hirm_io.load_observations(args.observations)
+  print(f"Loading clusters from {args.clusters} ...")
+  clusters = hirm_io.load_clusters(args.clusters)
+
+  print(f"Writing html to {args.output} ...")
+  visualize.plot_unary_matrices(clusters, obs, args.output)
+
+
+if __name__ == '__main__':
```
[the remark this answers is not in the document]
Comment (ThomasColthurst) [reply py/visualize_unary.py]: Done.

File: py/visualize_test.py [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+    observations.append(hirm_io.Observation(relation="R1", value="0", items=["A"]))
+    observations.append(hirm_io.Observation(relation="R1", value="1", items=["B"]))
+    observations.append(hirm_io.Observation(relation="R2", value="0", items=["A"]))
+    observations.append(hirm_io.Observation(relation="R2", value="1", items=["B"]))
+    m = visualize.make_numpy_matrix(observations, ["A", "B"], ["R1", "R2"])
+    self.assertEqual(m.shape, (2, 2))
+
+  def test_normalize_matrix(self):
+    m = np.identity(3)
+    np.testing.assert_equal(m, visualize.normalize_matrix(m))
```
[the remark this answers is not in the document]
Comment (ThomasColthurst) [reply py/visualize_test.py]: Done.

File: py/visualize.py [hunk 5]
```diff
@@ -0,0 +1,99 @@
...
+
+  # Matrix of data
+  m = make_numpy_matrix(obs, all_entities, all_relations)
+  cmap = copy.copy(plt.get_cmap('Greys'))
+  cmap.set_bad(color='gray')
+  ax.imshow(normalize_matrix(m))
+
+  # Red lines between columns (relations)
+  d = 0
+  for i in range(len(clusters)):
```
[the remark this answers is not in the document]
Comment (ThomasColthurst) [reply py/visualize.py]: Done.

File: py/hirm_io_test.py [hunk 6]
```diff
@@ -0,0 +1,29 @@
...
+    self.assertEqual(clusters[1].cluster_id, "3")
+    self.assertEqual(clusters[2].cluster_id, "6")
+    self.assertEqual(clusters[0].relations[0], "paws")
+    self.assertEqual(len(clusters[0].domain_clusters), 3)
+    self.assertEqual(clusters[0].domain_clusters[0].cluster_id, "0")
+    self.assertEqual(clusters[0].domain_clusters[0].domain, "animal")
+    self.assertEqual(clusters[0].domain_clusters[0].entities[0], "beaver")
+
+
+if __name__ == '__main__':
```
[the remark this answers is not in the document]
Comment (ThomasColthurst) [reply py/hirm_io_test.py]: Done.

File: py/visualize.py [hunk 7]
```diff
@@ -0,0 +1,99 @@
...
+    if cluster.cluster_id == clusters[i].cluster_id:
+      ax.axvline(d - 0.5, color='r', linewidth=2)
+    d += len(clusters[i].relations)
+    if cluster.cluster_id == clusters[i].cluster_id:
+      ax.axvline(d - 0.5, color='r', linewidth=2)
+      break
+    # if i > 0:
```
[the remark this answers is not in the document]
Comment (ThomasColthurst) [reply py/visualize.py]: Done.

**Comment by ThomasColthurst** [comment]
Thanks for the review!

**Pull request merged**